database/db.py

```diff
@@ -84,6 +84,37 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         return self.context.result
 
 
+class FailureException(Exception):
+    """An exception raised to give more information about the failure."""
```

Comment:
I debated about making this base class abstract, but decided to allow the creation of generic failure exceptions. That said, creating a subclass is useful for catching specific types of failures.